README.md
Bit6 Cordova Plugin
-------------------
[![GitHub version](https://badge.fury.io/gh/bit6%2Fbit6-cordova.svg)](https://github.com/bit6/bit6-cordova)

Add voice/video calling, texting, and rich media messaging into any mobile or web application.

Get Bit6 sample app running in 10 minutes - follow our Quick Start guides for:
* [Cordova CLI](http://docs.bit6.com/start/cordova-cli/)
* [Intel XDK](http://docs.bit6.com/start/cordova-xdk/)
* [Telerik App Builder](http://docs.bit6.com/start/cordova-telerik/)

### Prerequisites
* Get the API Key at [Bit6 Dashboard](https://dashboard.bit6.com).

### Installation
```sh
$ cordova plugin add https://github.com/bit6/bit6-cordova
```

### Configuration
Initialize Bit6 Plugin with your API key.
```js
document.addEventListener("deviceready", onDeviceReady, false);
function onDeviceReady() {
  var opts = {'apikey': 'yourApiKey'};
  var b6 = Bit6.init(opts);
  // Bit6 SDK has been initialized
}
```

### Video/Voice calls on iOS

To enable call support on iOS add [iosrtc](https://github.com/eface2face/cordova-plugin-iosrtc) plugin

```
cordova plugin add cordova-plugin-iosrtc
```

Note: Check the plugin requirements on [this page](https://github.com/eface2face/cordova-plugin-iosrtc) and [building page](https://github.com/eface2face/cordova-plugin-iosrtc/blob/master/docs/Building.md)

Note: no need to add iosrtc if your app does not use audio/video calls.


### Video/Voice calls on Android < 5

To enable call support on Android < 5 add [Crosswalk Webview](https://github.com/crosswalk-project/cordova-plugin-crosswalk-webview) plugin

```
cordova plugin add cordova-plugin-crosswalk-webview
```

### Real time notifications on Android<4.4

For real time notifications (websocket) on Android<4.4 Crosswalk Webview is required.

### Permissions on Android >= 6

For call support on Android >= 6 grant `Camera` and  `Microphone` permissions from `Settings->Apps->YourApp->Permissions`.

`Storage` permission is required to support message attachments.

### Quickstart
Create a user with a password authentication:
```js
b6.session.signup({identity: 'usr:john', password: 'secret'}, function(err) {
  if (err) {
    console.log('login error', err);
  } else {
    console.log('login successful');
  }
});
```
Send a message to another user:
```js
b6.compose('usr:tom').text('Hello!').send(function(err) {
  if (err) {
    console.log('error', err);
  } else {
    console.log('message sent');
  }
});
```
Make a video call:
```js
// Start an outgoing call and get a call controller (Dialog)
var d = b6.startCall('usr:tom', {audio: true, video: true});
```

### Documentation
Bit6 Cordova Plugin exposes the same API as [Bit6 JS SDK](https://github.com/bit6/bit6-js-sdk). Check Bit6 [JS documentation](http://docs.bit6.com/guides/js/).

### Demo app
The complete source code is available in the [demo repo](https://github.com/bit6/bit6-cordova-demo). Check out the same demo app running with JS SDK at http://demo.bit6.com.

### Supported platforms
* iOS
* Android
* Browser

### Push notifications

Push Notification support is required for receiving incoming calls and messages.

To enable this functionality please add one of these push plugins to your project:

* [phonegap-plugin-push](https://github.com/phonegap/phonegap-plugin-push) (recommended)
  ```sh
  # Set any value for SENDER_ID. Bit6 plugin will override it with the correct one.
  $ cordova plugin add phonegap-plugin-push --variable SENDER_ID="XXXXXXX"
  ```

* [Telerik Push Plugin](https://github.com/Telerik-Verified-Plugins/PushNotification)
  ```sh
  $ cordova plugin add https://github.com/Telerik-Verified-Plugins/PushNotification
  ```

* [Legacy PhoneGap Push Plugin](https://github.com/phonegap-build/PushPlugin) If you use cordova-ios 3.x


Then complete platform-specific configuration:

* __iOS APNs__ ([detailed guide](http://docs.bit6.com/guides/push-apns/))
    1. Generate APNS certificate in iTunes Connect.
    2. Export it into a p12 file.
    3. Add the file to your app in [Bit6 Dashboard](https://dashboard.bit6.com).

* __Android GCM__ ([detailed guide](http://docs.bit6.com/guides/push-gcm/))
    1. Get the project number and server key from [Google Dev Console](http://developer.android.com/google/gcm/gs.html).
    2. Add project number and server key for your app in [Bit6 Dashboard](https://dashboard.bit6.com).

### Building with Xcode 7
Please disable Bitcode support when building your Cordova app with Xcode.
Go to `Build Settings`, set `Enable Bitcode` to `No`.

### Third-party libraries
Bit6 plugin leverages code from the excellent [WebRTC](http://www.webrtc.org/), [PhoneRTC](https://github.com/alongubkin/phonertc) and [phonegap-websocket](https://github.com/mkuklis/phonegap-websocket/) projects.


CHANGELOG.md
## 0.11.0 [2017-05-05]

### Release Notes
Changed plugin name, added package.json, updated README

### Bugfixes
- Fixed the issue with sending video attachments for which the thumbnail generation fails.

## 0.10.0 [2016-12-23]

### Release Notes
Phonertc module is removed. For Webrtc support on iOS and on Adnroid<5 iosrtc and Crosswalk webview plugins are used.


### Breaking Changes
- For voice/video calls (iOS, Android<5) the plugins mentioned above are required.
- For real time notifications support on Android<4.4 Crosswalk Webview is required.

### Bugfixes
- Fixed the issue with missing remote video during iOS calls
- iOS 10 support

## 0.9.9 [2016-09-02]

### Release Notes

This release adds support for camera selection for video calls and audio recording during the calls.

### Features

- Added support for camera selection on initiating a video call
- Added support for call recording

### Bugfixes

- Fixed the issue with push key cleanup. Now after logout the user will not get push notifications.


## 0.9.8 [2016-05-25]

### Release Notes

This release adds support for custom push handling, switchCamera api and a mechanism to keep the user logged in.

### Features

- Added support for custom push handling/implementation.
- Added switchCamera method to Bit6 api: supported only on Android < 5.0
- Added a mechanism to keep the user logged in after app restart.

### Bugfixes

- Fixed the issue with opening the app by tapping on a notification when the app is closed.


## 0.9.7 [2016-04-28]

### Release Notes

This release adds support for new push plugin (phonegap-plugin-push) while keeping the compatibility with the previously supported push plugins.

### Features

- Added support for phonegap-plugin-push
- Added push alerts localizatoin support on iOS. Works with new phonegap-plugin-push only

### Bugfixes

- Fixed the call issue appearing on some Android devices caused by a missing permission
- Fixed issues with Intel XDK Emulator (iOS, Android) and debugging in AppPreview (Android only)
- Fixed camera selection on Android to use the front camera for video calls

## 0.9.6.1 [2016-03-15]

### Release Notes

This release adds support for cordova-ios 4.x, reduces plugin size, and fixes an issue with iOS push notifications for Ad Hoc builds.

### Breaking Changes

- iOS simulator build is not supported - i386 slice is removed from iOS WebRTC library
- Push plugin dependency must be specified manually in the application - see 'Push notifications' section in README

### Features

- Added support for cordova-ios 4.x
- Significantly reduced plugin size
- Updated WebRTC lib: Android - m49

### Bugfixes

- Fixed iOS push issue for Ad Hoc builds


## 0.9.6 [2016-02-11]

### Release Notes

This release has improved API for video elements, updated native WebRTC libs and some important bug fixes.

### Features

- New video element layout mechanism
- External authentication support
- Automatic selection of iOS APNS environment
- Updated WebRTC libs : iOS - m49, Andoid - m48

### Bugfixes

- Fixed the call setup issue between devices (caused by incorrect buffering of ICE candidates)
- Fixed the whitelist plugin warning about security policy tag


www/phonertc.js
var exec = require('cordova/exec');
var videoViewConfig;

function createUUID() {
  // http://www.ietf.org/rfc/rfc4122.txt
  var s = [];
  var hexDigits = "0123456789abcdef";
  for (var i = 0; i < 36; i++) {
      s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1);
  }
  s[14] = "4";  // bits 12-15 of the time_hi_and_version field to 0010
  s[19] = hexDigits.substr((s[19] & 0x3) | 0x8, 1);  // bits 6-7 of the clock_seq_hi_and_reserved to 01
  s[8] = s[13] = s[18] = s[23] = "-";

  var uuid = s.join("");
  return uuid;
}

function Session(config) {
  // make sure that the config object is valid
  if (typeof config !== 'object') {
    throw {
      name: 'PhoneRTC Error',
      message: 'The first argument must be an object.'
    };
  }

  //Small hack to make the original config parser work
  if (config.streams.video) {
    var streams = config.streams;
    streams.video = true;//Overriding  {video:{facingMode: user}} format
    config.streams = streams;
  }

  if (typeof config.isInitiator === 'undefined' ||
      typeof config.turn === 'undefined' ||
      typeof config.streams === 'undefined') {
    throw {
      name: 'PhoneRTC Error',
      message: 'isInitiator, turn and streams are required parameters.'
    };
  }

  var self = this;
  self.events = {};
  self.config = config;
  self.sessionKey = createUUID();

  // make all config properties accessible from this object
  Object.keys(config).forEach(function (prop) {
    Object.defineProperty(self, prop, {
      get: function () { return self.config[prop]; },
      set: function (value) { self.config[prop] = value; }
    });
  });

  function callEvent(eventName) {
    if (!self.events[eventName]) {
      return;
    }

    var args = Array.prototype.slice.call(arguments, 1);
    self.events[eventName].forEach(function (callback) {
      callback.apply(self, args);
    });
  }

  function onSendMessage(data) {
    if (data.type === '__answered') {
      callEvent('answer');
    } else if (data.type === '__disconnected') {
      callEvent('disconnect');
    } else {
      callEvent('sendMessage', data);
    }
  }

  exec(onSendMessage, null, 'PhoneRTCPlugin', 'createSessionObject', [self.sessionKey, config]);
};

Session.prototype.on = function (eventName, fn) {
  // make sure that the second argument is a function
  if (typeof fn !== 'function') {
    throw {
      name: 'PhoneRTC Error',
      message: 'The second argument must be a function.'
    };
  }

  // create the event if it doesn't exist
  if (!this.events[eventName]) {
    this.events[eventName] = [];
  } else {
    // make sure that this callback doesn't exist already
    for (var i = 0, len = this.events[eventName].length; i < len; i++) {
      if (this.events[eventName][i] === fn) {
        throw {
          name: 'PhoneRTC Error',
          message: 'This callback function was already added.'
        };
      }
    }
  }

  // add the event
  this.events[eventName].push(fn);
};

Session.prototype.off = function (eventName, fn) {
  // make sure that the second argument is a function
  if (typeof fn !== 'function') {
    throw {
      name: 'PhoneRTC Error',
      message: 'The second argument must be a function.'
    };
  }

  if (!this.events[eventName]) {
    return;
  }

  var indexesToRemove = [];
  for (var i = 0, len = this.events[eventName].length; i < len; i++) {
    if (this.events[eventName][i] === fn) {
      indexesToRemove.push(i);
    }
  }

  indexesToRemove.forEach(function (index) {
    this.events.splice(index, 1);
  })
};

Session.prototype.call = function (success, error) {
  exec(success, error, 'PhoneRTCPlugin', 'call', [{
    sessionKey: this.sessionKey
  }]);
};

Session.prototype.receiveMessage = function (data) {
  exec(null, null, 'PhoneRTCPlugin', 'receiveMessage', [{
    sessionKey: this.sessionKey,
    message: JSON.stringify(data)
  }]);
};

Session.prototype.renegotiate = function () {
  exec(null, null, 'PhoneRTCPlugin', 'renegotiate', [{
    sessionKey: this.sessionKey,
    config: this.config
  }]);
};

Session.prototype.close = function () {
  exec(null, null, 'PhoneRTCPlugin', 'disconnect', [{
    sessionKey: this.sessionKey
  }]);
};

exports.Session = Session;

function getLayoutParams(videoElement) {
  var boundingRect = videoElement.getBoundingClientRect();

  if (cordova.platformId === 'android') {
    return {
      position: [boundingRect.left + window.scrollX, boundingRect.top + window.scrollY],
      size: [videoElement.offsetWidth, videoElement.offsetHeight]
    };
  }

  return {
    position: [boundingRect.left, boundingRect.top],
    size: [videoElement.offsetWidth, videoElement.offsetHeight]
  };
}

function setVideoView(config) {
  videoViewConfig = config;

  var container = config.container;

  if (container) {
    config.containerParams = getLayoutParams(container);
    delete config.container;
  }

  config.devicePixelRatio = window.devicePixelRatio || 2;

  exec(null, null, 'PhoneRTCPlugin', 'setVideoView', [config]);

  if (container) {
    config.container = container;
  }
};

document.addEventListener('touchmove', function () {
  if (videoViewConfig) {
    setVideoView(videoViewConfig);
  }
});

exports.setVideoView = setVideoView;

exports.hideVideoView = function () {
  exec(null, null, 'PhoneRTCPlugin', 'hideVideoView', []);
};

exports.showVideoView = function () {
  exec(null, null, 'PhoneRTCPlugin', 'showVideoView', []);
};

function isApnsProduction(onSuccess) {
  if (cordova.platformId === 'ios') { //This is needed only for ios.
    exec(onSuccess, null, 'PhoneRTCPlugin', 'isApnsProduction', []);
  }
  else {
    onSuccess(false);
  }
}

exports.isApnsProduction = isApnsProduction;

function switchCamera() {
  exec(null, null, 'PhoneRTCPlugin', 'switchCamera', []);
}
exports.switchCamera = switchCamera;

function selectCamera(videoConf) {
  var isFrontCam = (videoConf != null);
  if (videoConf && videoConf.facingMode == "environment") {
    isFrontCam = false;
  }

  exec(null, null, 'PhoneRTCPlugin', 'selectCamera', [isFrontCam]);
}

exports.selectCamera = selectCamera;

